### PR TITLE
RS/CH/Rule6-5

### PR DESCRIPTION
**Pull request by charliepnnl**
The PR also include a new utility function, *invert_mask*.

Note: Black modified several files.

Although I removed the thermal_zone loops from the RDS, I did not modify the calls to *normalize_interior_lighting_schedules*. Ideally, this would be updated also.

### CODE REVIEW

File: rct229/rules/section6/section6rule5.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                super(Section6Rule5.BuildingRule.ZoneRule, self).__init__(
+                    rmrs_used=UserBaselineProposedVals(False, True, True),
+                    each_rule=Section6Rule5.BuildingRule.ZoneRule.SpaceRule(),
+                    index_rmr="baseline",
+                )
+
+            def create_data(self, context, data=None):
+                return {
+                    "avg_zone_height_b": get_avg_zone_height(zone_b),
+                    "avg_zone_height_p": get_avg_zone_height(zone_p),
```
Comment (yunjoonjung-PNNL) [inline rct229/rules/section6/section6rule5.py:63]: Please kindly ignore if I'm wrong, Charlie. 
Shouldn't ```zone_p``` and ```zone_b``` be defined before (e.g., zone_ b = context.baseline in line 61)?

**Review by yunjoonjung-PNNL — APPROVED** [review]
Thanks for doing this task, Charlie. I'd appreciate it if you could check my one comment.

File: rct229/rules/section6/section6rule5.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
+from rct229.rule_engine.rule_base import RuleDefinitionBase
+from rct229.rule_engine.rule_list_indexed_base import RuleDefinitionListIndexedBase
+from rct229.rule_engine.user_baseline_proposed_vals import UserBaselineProposedVals
+from rct229.schema.config import ureg
+from rct229.utils.masks import invert_mask
+
```
Comment (yunjoonjung-PNNL) [inline rct229/rules/section6/section6rule5.py:6]: I believe ```pint_sum```, ```get_avg_zone_height``` ```ZERO```, ```normalize_interior_lighting_schedules```, ```compare_schedules``` function are not imported. Could you verify these?

**Review by yunjoonjung-PNNL — APPROVED** [review]

File: rct229/rules/section6/section6rule5.py (hunk 1 above)
Comment (charliepnnl) [reply rct229/rules/section6/section6rule5.py:63]: Yes, thank you.

File: rct229/rules/section6/section6rule5.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                required_fields={"$": ["building_open_schedule"]},
+                data_items={
+                    "building_open_schedule_b",
+                    ("baseline", "building_open_schedule"),
+                },
+            )
+
+        def is_applicable(self, context, data):
+            building_total_area_b = pint_sum(
+                "$..spaces[*].floor_area", building_b, ZERO.AREA
```
Comment (weilixu) [inline rct229/rules/section6/section6rule5.py:47]: We will need to add a "find_all" here.
```python
building_b = context.baseline
pint_sum(find_all("$..spaces[*].floor_area", building_b), ZERO.AREA)
```

File: rct229/rules/section6/section6rule5.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            def create_data(self, context, data=None):
+                return {
+                    "avg_zone_height_b": get_avg_zone_height(zone_b),
+                    "avg_zone_height_p": get_avg_zone_height(zone_p),
+                }
+
+            class SpaceRule(RuleDefinitionBase):
+                def __init__(self):
+                    super(Section6Rule5.BuildingRule.ZoneRule.SpaceRule, self).__init__(
+                        rmrs_used=UserBaselineProposedVals(True, False, True)
```
Comment (weilixu) [inline rct229/rules/section6/section6rule5.py:69]: This should be False, True, True.

File: rct229/rules/section6/section6rule5.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                        is_leap_year=is_leap_year_b,
+                    )
+
+                    return {"schedule_comparison_result": schedule_comparison_result}
+
+                def rule_check(self, context, calc_vals=None, data=None):
+                    schedule_comparison_result = calc_vals["schedule_comparison_result"]
+
+                    return (
+                        schedule_comparison_result["total+hours_matched"]
```
Comment (weilixu) [inline rct229/rules/section6/section6rule5.py:110]: typo "total_hours_matched"

**Review by weilixu — CHANGES_REQUESTED** [review]
Request updates - but also need to wait for RDS reply on the issue.

File: rct229/rules/section6/section6rule5.py (hunk 2 above)
Comment (charliepnnl) [reply rct229/rules/section6/section6rule5.py:6]: Yes, thank you.

File: rct229/rules/section6/section6rule5.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    class BuildingRule(RuleDefinitionListIndexedBase):
+        def __init__(self):
+            super(Section6Rule5.BuildingRule, self).__init__(
+                rmrs_used=UserBaselineProposedVals(False, True, True),
+                each_rule=Section6Rule5.BuildingRule.ZoneRule(),
+                index_rmr="baseline",
+                required_fields={"$": ["building_open_schedule"]},
+                data_items={
+                    "building_open_schedule_b",
+                    ("baseline", "building_open_schedule"),
```
[the remark this answers is not in the document]
Comment (charliepnnl) [reply rct229/rules/section6/section6rule5.py:41]: You are correct. Thanks.

File: rct229/rules/section6/section6rule5.py (hunk 3 above)
Comment (charliepnnl) [reply rct229/rules/section6/section6rule5.py:47]: Yep. Thanks.

File: rct229/rules/section6/section6rule5.py (hunk 4 above)
Comment (charliepnnl) [reply rct229/rules/section6/section6rule5.py:69]: Thank you.

File: rct229/rules/section6/section6rule5.py (hunk 5 above)
Comment (charliepnnl) [reply rct229/rules/section6/section6rule5.py:110]: Yes. Thanks.

**Comment by charliepnnl** [comment]
I have made all the suggested changes. 
Thank you Weili and Yun Joon for catching all these mistakes.

**Comment by yunjoonjung-PNNL** [comment]
I just tested rule 6-5 with the testing JSON file. Could you check the followings?

1. line 27: required_fields: ```schedules``` key is under the ```ruleset_model_instances``` key level. Could you double-check this?
2. line 54: find_all: I think ```ZERO.AREA``` should be outside the ```find_all``` function (currently, there are 3 arguments instead of 2).
3. line 71: ```zone_b``` and ```zone_p``` are on the building level. So, using ```get_avg_zone_height``` caused an error because the level isn't on zone level.

I'd appreciate it if you could verify these. Thank you!

**Comment by charliepnnl** [comment]
> 
@yunjoonjung-PNNL I fixed your issues 1 and 2.  I'm not seeing your issue 3.  I included the test json you have me and I now get a schema error for the proposed building:  
```
INVALID SCHEMA: Test rule-6-5-a: Proposed RMR: schema invalid: None is not of type 'object'
{'invalid_rmrs': {'Proposed': "schema invalid: None is not of type 'object'"},
 'outcomes': []}
INVALID SCHEMA: Test rule-6-5-b: Proposed RMR: schema invalid: None is not of type 'object'
{'invalid_rmrs': {'Proposed': "schema invalid: None is not of type 'object'"},
 'outcomes': []}
INVALID SCHEMA: Test rule-6-5-c: Proposed RMR: schema invalid: None is not of type 'object'
```

**Comment by weilixu** [comment]
> I just tested rule 6-5 with the testing JSON file. Could you check the followings?
> 
> 1. line 27: required_fields: `schedules` key is under the `ruleset_model_instances` key level. Could you double-check this?
> 2. line 54: find_all: I think `ZERO.AREA` should be outside the `find_all` function (currently, there are 3 arguments instead of 2).
> 3. line 71: `zone_b` and `zone_p` are on the building level. So, using `get_avg_zone_height` caused an error because the level isn't on zone level.
> 
> I'd appreciate it if you could verify these. Thank you!

I did not see line 71 has `zone_b` and `zone_p`, does it updated?

File: rct229/rules/section6/section6rule5.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+            id="6-5",
+            rmrs_used=UserBaselineProposedVals(False, True, True),
+            each_rule=Section6Rule5.RulesetModelInstanceRule(),
+            index_rmr="baseline",
+            description="Baseline building is modeled with automatic shutoff controls in buildings >5000 sq.ft.",
+            required_fields={
+                "$": ["calendar"],
+                "calendar": ["is_leap_year"],
+            },
+            list_path="ruleset_model_instances[0].buildings[*]",
```
Comment (weilixu) [inline rct229/rules/section6/section6rule5.py:31]: `list_path="ruleset_model_instances[*]`

Adding building will point the nested class to building level.
In this case, the search for schedules will be in buildings not in RMI - which causes validation error.

File: rct229/rules/section6/section6rule5.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                    each_rule=Section6Rule5.RulesetModelInstanceRule.BuildingRule.ZoneRule(),
+                    index_rmr="baseline",
+                    required_fields={"$": ["building_open_schedule"]},
+                    data_items={
+                        "building_open_schedule_b": (
+                            "baseline",
+                            "building_open_schedule",
+                        ),
+                    },
+                )
```
Comment (weilixu) [inline rct229/rules/section6/section6rule5.py:64]: Building rule is missing list_path points to the zone.

add 
`list_path="building_segments[*].zones[*]",` in the initiator

File: rct229/rules/section6/section6rule5.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+            class ZoneRule(RuleDefinitionListIndexedBase):
+                def __init__(self):
+                    super(
+                        Section6Rule5.RulesetModelInstanceRule.BuildingRule.ZoneRule,
+                        self,
+                    ).__init__(
+                        rmrs_used=UserBaselineProposedVals(False, True, True),
+                        each_rule=Section6Rule5.RulesetModelInstanceRule.BuildingRule.ZoneRule.SpaceRule(),
+                        index_rmr="baseline",
```
Comment (weilixu) [inline rct229/rules/section6/section6rule5.py:82]: Zone rule is missing list_path as well.
Add

`list_path="spaces[*]"

File: rct229/rules/section6/section6rule5.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                                space_height_p,
+                                schedules_p,
+                                adjust_for_credit=False,
+                            )
+                        )
+
+                        schedule_comparison_result = compare_schedules(
+                            normalized_interior_lighting_schedule_b,
+                            normalized_interior_lighting_schedule_p,
+                            mask_schedule=invert_mask(building_open_schedule_b),
```
Comment (weilixu) [inline rct229/rules/section6/section6rule5.py:130]: The `building_open_schedule_b` is not declared in this function.

**Review by weilixu — COMMENTED** [review]
Thanks for addressing the comments - There are more issues need to be addressed both in code and test file.
See comments for issues in code
test json:
1. 6-5-c is missing schedules in RMI.
2. All spaces in test cases are missing `lighting_space_type`.

**Comment by charliepnnl** [comment]
I have made the changes to the Rule based on Weili's comments.

**Comment by weilixu** [comment]
> I have made the changes to the Rule based on Weili's comments.

Thanks! I think the remaining is updating the json test file and waiting for RDS team to interpret the assertions.

**Review by weilixu — APPROVED** [review]

**Comment by weilixu** [comment]
@charliepnnl The issue is answered and the 6-7 RDS is updated #567 . Could you update the code to reflect the latest change?

**Comment by charliepnnl** [comment]
> @charliepnnl The issue is answered and the 6-7 RDS is updated #567 . Could you update the code to reflect the latest change?

@weilixu Did you mean to put this comment in Yun Joon's PR#503 for Rule6-7?

**Comment by charliepnnl** [comment]
I have fixed all issues from comments and all json tests run as expected.

**Comment by charliepnnl** [comment]
Note:  I also updated the RDS to reflect how the schedules actually appear in the RDS.

**Pull request merged**